inc/Engine/CriticalPath/RESTGenerate.php

```diff
@@ -0,0 +1,302 @@
+	 *
+	 * @since 3.6
+	 *
+	 * @return void
+	 */
+	public function register_generate_route() {
+		register_rest_route(
+			self::ROUTE_NAMESPACE,
+			'cpcss/post/(?P<id>[\d]+)',
+			[
+				'methods'             => 'POST',
+				'callback'            => [ $this, 'generate' ],
+				'permission_callback' => [ $this, 'check_permissions' ],
+			]
+		);
+	}
+
+	/**
+	 * Generates the CPCSS for the requested post ID
+	 *
+	 * @since 3.6
+	 *
+	 * @param WP_REST_Request $request WP REST request response.
+	 * @return WP_REST_Response
+	 */
+	public function generate( WP_REST_Request $request ) {
+		$status = get_post_status( $request['id'] );
+
+		if ( ! $status ) {
+			return rest_ensure_response(
+				[
+					'success' => false,
+					'code'    => 'post_not_exists',
+					'message' => __( 'Requested post does not exist.', 'rocket' ),
+					'data'    => [
+						'status' => 400,
+					],
+				]
+			);
+		}
+
+		if ( 'publish' !== $status ) {
+			return rest_ensure_response(
+				[
+					'success' => false,
+					'code'    => 'post_not_published',
+					'message' => __( 'Cannot generate CPCSS for unpublished post.', 'rocket' ),
+					'data'    => [
+						'status' => 400,
+					],
+				]
+			);
+		}
+
+		$post_url  = get_permalink( $request['id'] );
+		$post_type = get_post_type( $request['id'] );
+
+		$response = $this->send_generation_request( $post_url, $post_type );
+
+		if ( false === $response['success'] ) {
+			return rest_ensure_response( $response );
+		}
+
+		while ( $job_data = $this->get_critical_path( $response['data']['id'] ) ) { // phpcs:ignore WordPress.CodeAnalysis.AssignmentInCondition.FoundInWhileCondition
+			if ( in_array( (int) $job_data->status, [ 400, 404 ], true ) ) {
+				// translators: %1$s = post URL.
+				$error = sprintf( __( 'Critical CSS for %1$s not generated.', 'rocket' ), $post_url );
+
+				if ( isset( $job_data->message ) ) {
+					// translators: %1$s = error message.
+					$error .= ' ' . sprintf( __( 'Error: %1$s', 'rocket' ), $job_data->message );
+				}
+
+				return rest_ensure_response(
+					[
+						'success' => false,
+						'code'    => 'cpcss_generation_failed',
+						'message' => $error,
+						'data'    => [
+							'status' => 400,
+						],
+					]
+				);
+			}
+
+			if ( isset( $job_data->data->state, $job_data->data->critical_path ) && 'complete' === $job_data->data->state ) {
+				if ( ! $this->save_post_cpcss( $request['id'], $post_type, $job_data->data->critical_path ) ) {
+					return rest_ensure_response(
+						[
+							'success' => false,
+							'code'    => 'cpcss_generation_failed',
+							'message' => sprintf(
+								// translators: %1$s = post URL, %2$s = error message.
+								__( 'Critical CSS for %1$s not generated. Error: %2$s', 'rocket' ),
+								$post_url,
+								// translators: %s = critical CSS directory path.
+								sprintf( __( 'The critical CSS content could not be saved as a file in %s.', 'rocket' ), $this->critical_css_path )
+							),
```

Comment:
Since the first sentence seems to be used only once, it may be simpler to do this:
```suggestion
							'message' => sprintf(
								// translators: %1$s = post URL, %2$s = critical CSS directory path.
								__( 'Critical CSS for %1$s not generated. Error: The critical CSS content could not be saved as a file in %2$s', 'rocket' ),
								$post_url,
								$this->critical_css_path
							),
```
But I think I understand why you did this that way. What do you think?

Reply:
I don't know why I split those like those, your suggestion makes sense too